private/bufpkg/bufimage/bufimageutil/bufimageutil.go

```diff
@@ -60,6 +61,470 @@ func FreeMessageRangeStrings(
+		i = 0
+		for _, serviceDescriptor := range imageFileDescriptor.Service {
+			name := prefix + serviceDescriptor.GetName()
+			if _, ok := typesToKeep[name]; ok {
+				imageFileDescriptor.Service[i] = serviceDescriptor
+				i++
+			}
+		}
+		imageFileDescriptor.Service = imageFileDescriptor.Service[:i]
+
+		// With some from/to mappings, perhaps even sourcecodeinfo isn't too bad.
+		imageFileDescriptor.SourceCodeInfo = nil
+	}
+	return bufimage.NewImage(includedFiles)
+}
+
+// trimMessageDescriptor removes (nested) messages and nested enums from a slice
+// of message descriptors if their type names are not found in the toKeep map.
+func trimMessageDescriptor(in []*descriptorpb.DescriptorProto, prefix string, toKeep map[string]struct{}) ([]*descriptorpb.DescriptorProto, error) {
+	i := 0
+	for _, messageDescriptor := range in {
+		name := prefix + messageDescriptor.GetName()
+		if _, ok := toKeep[name]; ok {
+			trimMessages, err := trimMessageDescriptor(messageDescriptor.NestedType, name+".", toKeep)
+			if err != nil {
+				return nil, err
+			}
+			messageDescriptor.NestedType = trimMessages
+			trimEnums, err := trimEnumDescriptor(messageDescriptor.EnumType, name+".", toKeep)
+			if err != nil {
+				return nil, err
+			}
+			messageDescriptor.EnumType = trimEnums
+			trimExtensions, err := trimExtensionDescriptors(messageDescriptor.Extension, name+".", toKeep)
+			if err != nil {
+				return nil, err
+			}
+			messageDescriptor.Extension = trimExtensions
+			in[i] = messageDescriptor
+			i++
+		}
+	}
+	return in[:i], nil
+}
+
+// trimEnumDescriptor removes enums from a slice of enum descriptors if their
+// type names are not found in the toKeep map.
+func trimEnumDescriptor(in []*descriptorpb.EnumDescriptorProto, prefix string, toKeep map[string]struct{}) ([]*descriptorpb.EnumDescriptorProto, error) {
+	i := 0
+	for _, enumDescriptor := range in {
+		name := prefix + enumDescriptor.GetName()
+		if _, ok := toKeep[name]; ok {
+			in[i] = enumDescriptor
+			i++
+		}
+	}
+	return in[:i], nil
+}
+
+func trimExtensionDescriptors(in []*descriptorpb.FieldDescriptorProto, prefix string, toKeep map[string]struct{}) ([]*descriptorpb.FieldDescriptorProto, error) {
+	i := 0
+	for _, fieldDescriptor := range in {
+		name := prefix + fieldDescriptor.GetName()
+		if _, ok := toKeep[name]; ok {
+			in[i] = fieldDescriptor
+			i++
+		}
+	}
+	return in[:i], nil
+}
+
+// descriptorAndDirects holds a protsource.NamedDescriptor and a list of all
+// named descriptors it directly references. A directly referenced dependency is
+// any type that if defined in a different file from the principal descriptor,
+// an import statement would be required for the proto to compile.
+type descriptorAndDirects struct {
+	Descriptor protosource.NamedDescriptor
+	Directs    []protosource.NamedDescriptor
+}
+
+func descriptorTransitiveClosure(namedDescriptor protosource.NamedDescriptor, imageIndex *imageIndex, seen map[string]struct{}) ([]descriptorAndDirects, error) {
```

Comment:
Overall the idea is sane. I trust that this covers all the bases and is sufficiently DRY'd up.